others/freivalds_algorithm.cpp

```diff
@@ -0,0 +1,169 @@
+/**
+ * @file
+ * @brief Implementation of [Freivalds' algorithm]
+ * (https://en.wikipedia.org/wiki/Freivalds%27_algorithm)
+ * @author [Darshana Sarma](https://github.com/Darshana-Sarma)
+ * @details
+ * Given three n × n matrices A, B, and
+ * C, a general problem is to verify whether AB=C
+ *
+ * The time Complexity of the algorithm is O(n^{2}).
+ */
+#include <array>     //   for std::array
+#include <cassert>   //   for assert
+#include <iostream>  //   for io operations
+
+/**
+ * @namespace others
+ * @brief Other algorithms
+ */
+namespace others {
+/**
+ * @namespace freivalds
+ * @brief Functions for [Freivalds']
+ * (https://en.wikipedia.org/wiki/Freivalds%27_algorithm) algorithm
+ */
+namespace freivalds {
+/**
+ * The function will check AB=C by applying Freidvald's algorithm 
+ * @tparam N dimension of square matrix NxN
+ * @param A square matrix A
+ * @param B square matrix B
+ * @param C resultant square matrix C
+ * @returns boolean value
+ */
+template <std::size_t N>
+bool CoppersmithAlgo(std::array<std::array<double, N>, N> A,
+                     std::array<std::array<double, N>, N> B,
+                     std::array<std::array<double, N>, N> C) {
+    // Printing matrix A
+    std::cout << "\nThe first matrix is:\n ";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            std::cout << A[i][j] << "\t";
+        }
+        std::cout << "\n";
+    }
+
+    // Printing matrix B
+    std::cout << "\nThe Second matrix is:\n ";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            std::cout << B[i][j] << "\t";
+        }
+        std::cout << "\n";
+    }
+
+    // Printing matrix C
+    std::cout << "The resultant matrix: \n";
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            std::cout << C[i][j] << "\t";
+        }
+        std::cout << "\n";
+    }
+    // generate a random nx1 vector r containing 0/1 elements
+    std::array<std::array<const double, N>, 1> a{0};

```

Comment:
It seems like a rather bad practice to rely on generating random number from an arbitrary garbage in the memory. There's a good chance you'll come across an empty piece of memory and obtain a vector full of zeros. I suggest you use a proper (pseudo)-random number generator.

Reply:
I don't think this has anything to do with random number generators.
`a` is simple a vector/array = 1D matrix. 
What I understood was that the initialization `a{0}` will initialize the values of all its elements as 0 as:
1. Initialize the 1st dimension values to 0
2. implicitly applies initialize the elements of `a[0]` with 0.

However, this is counterintuitive to the use of the variable `a` on line 73 and 82.

Reply:
From the algorithm description and from the above comment, it is clear that the desired behaviour is to actually produce an array of N random numbers here.

Initialisation by `{0}` is very non-intuitive construct that I think should be avoided completely. If the intention is to have default value initialisation, it can be left out completely with the default constructor. I'm afraid, that according to the [specs](https://en.cppreference.com/w/cpp/container/array) this way when you only initialise the first element, you run into an issue with indeterminate values.

Reply:
Isn't this aggregate initialization: https://en.cppreference.com/w/cpp/language/aggregate_initialization